### ISSUE TITLE
$removeparam has problems/conflicts(?), not removing some parameters.

### ISSUE THREAD

**Issue by Emi-TheDhamphirInLoveUnderTheFrozenStar** [brave/adblock-rust#297]
Hey @antonok-edm 

Again here reporting, just an issue I noticed with the $removeparam feature.

It has to do with some parameters not getting removed, so hope this helps to improve it.

First, Brave doesn't seem to remove parameters if they are blank, example if you try to remove `?param1=&param2=test` only param2 will be removed. Not a big deal since it is a blank parameter but throwing it there in case you can improve that so parameters can be removed 100%.

Second and the reason for this issue is simple, Brave is having problems with parameters that have underscore on them, when there are too many it has troubles, removing what it should easily remove, because in small groups it seems to work, but once we have many rules that are like `parameter_likethis` problems happen.

It doesn't happen on every parameter, but it makes it imperfect, I tested removing all underscores from all the rules and test it in a dummy URL, and it removed all parameters perfectly, I finally noticed the common factor was the underscores for this weird problem so hope it can get fixed.

How to test:
Ad d these rules as custom filters, they are taken the Adguard param filter list, adding the $document as stated in my other issue to make them work in Brave.

```
$document,removeparam=famad_xuid
$document,removeparam=twclid
$document,removeparam=erid
$document,removeparam=cx_click
$document,removeparam=mkt_tok
$document,removeparam=mindbox-message-key
$document,removeparam=s_cid
$document,removeparam=awc
$document,removeparam=hsa_acc
$document,removeparam=hsa_ad
$document,removeparam=hsa_cam
$document,removeparam=hsa_grp
$document,removeparam=hsa_kw
$document,removeparam=hsa_la
$document,removeparam=hsa_mt
$document,removeparam=hsa_net
$document,removeparam=hsa_ol
$document,removeparam=hsa_src
$document,removeparam=hsa_tgt
$document,removeparam=hsa_ver
$document,removeparam=hsCtaTracking
$document,removeparam=ysclid
$document,removeparam=yclid
$document,removeparam=aiad_clid
$document,removeparam=mc_eid
$document,removeparam=maf
$document,removeparam=wt_mc
$document,removeparam=oprtrack
$document,removeparam=xtor
$document,removeparam=msclkid
$document,removeparam=vero_conv
$document,removeparam=vero_id
$document,removeparam=itm_source
$document,removeparam=itm_medium
$document,removeparam=itm_campaign
$document,removeparam=itm_content
$document,removeparam=itm_term
$document,removeparam=utm_ad
$document,removeparam=utm_affiliate
$document,removeparam=utm_brand
$document,removeparam=utm_campaign
$document,removeparam=utm_campaignid
$document,removeparam=utm_channel
$document,removeparam=utm_cid
$document,removeparam=utm_content
$document,removeparam=utm_creative
$document,removeparam=utm_emcid
$document,removeparam=utm_emmid
$document,removeparam=utm_id
$document,removeparam=utm_keyword
$document,removeparam=utm_medium
$document,removeparam=utm_name
$document,removeparam=utm_place
$document,removeparam=utm_product
$document,removeparam=utm_pubreferrer
$document,removeparam=utm_reader
$document,removeparam=utm_referrer
$document,removeparam=utm_serial
$document,removeparam=utm_session
$document,removeparam=utm_siteid
$document,removeparam=utm_social
$document,removeparam=utm_social-type
$document,removeparam=utm_source
$document,removeparam=utm_supplier
$document,removeparam=utm_swu
$document,removeparam=utm_term
$document,removeparam=utm_umguk
$document,removeparam=utm_userid
$document,removeparam=utm_viz_id
$document,removeparam=gbraid
$document,removeparam=wbraid
$document,removeparam=gclsrc
$document,removeparam=gclid
$document,removeparam=usqp
$document,removeparam=dpg_source
$document,removeparam=dpg_campaign
$document,removeparam=dpg_medium
$document,removeparam=dpg_content
$document,removeparam=admitad_uid
$document,removeparam=gps_adid
$document,removeparam=unicorn_click_id
$document,removeparam=adjust_creative
$document,removeparam=adjust_tracker_limit
$document,removeparam=adjust_tracker
$document,removeparam=adjust_adgroup
$document,removeparam=adjust_campaign
$document,removeparam=bsft_clkid
$document,removeparam=bsft_eid
$document,removeparam=bsft_mid
$document,removeparam=bsft_uid
$document,removeparam=bsft_aaid
$document,removeparam=bsft_ek
$document,removeparam=mtm_campaign
$document,removeparam=mtm_cid
$document,removeparam=mtm_content
$document,removeparam=mtm_group
$document,removeparam=mtm_keyword
$document,removeparam=mtm_medium
$document,removeparam=mtm_placement
$document,removeparam=mtm_source
$document,removeparam=pk_campaign
$document,removeparam=pk_medium
$document,removeparam=pk_source
$document,removeparam=vc_lpp
$document,removeparam=ml_subscriber
$document,removeparam=ml_subscriber_hash
$document,removeparam=rb_clickid
$document,removeparam=oly_anon_id
$document,removeparam=oly_enc_id
$document,removeparam=ebisAdID
$document,removeparam=wickedid
$document,removeparam=irgwc
$document,removeparam=fbclid
$document,removeparam=adfrom
$document,removeparam=nx_source
$document,removeparam=cmpid
$document,removeparam=asgtbndr
$document,removeparam=guccounter
$document,removeparam=guce_referrer
$document,removeparam=guce_referrer_sig
$document,removeparam=action_object_map
$document,removeparam=action_ref_map
$document,removeparam=action_type_map
$document,removeparam=fb_action_ids
$document,removeparam=fb_action_types
$document,removeparam=fb_comment_id
$document,removeparam=fb_ref
$document,removeparam=fb_source
$document,removeparam=dclid
```

Now even in a simple simple URL like `brave.com/?utm_medium=test&mtm_medium=test` you will notice that utm_medium=test parameter was not removed.

But if we make all parameters appear in the URL:

`https://brave.com/?famad_xuid=test&twclid=test&erid=test&cx_click=test&mkt_tok=test&mindbox-message-key=test&s_cid=test&awc=test&hsa_acc=test&hsa_ad=test&hsa_cam=test&hsa_grp=test&hsa_kw=test&hsa_la=test&hsa_mt=test&hsa_net=test&hsa_ol=test&hsa_src=test&hsa_tgt=test&hsa_ver=test&hsCtaTracking=test&ysclid=test&yclid=test&aiad_clid=test&mc_eid=test&maf=test&wt_mc=test&oprtrack=test&xtor=test&msclkid=test&vero_conv=test&vero_id=test&itm_source=test&itm_medium=test&itm_campaign=test&itm_content=test&itm_term=test&utm_ad=test&utm_affiliate=test&utm_brand=test&utm_campaign=test&utm_campaignid=test&utm_channel=test&utm_cid=test&utm_content=test&utm_creative=test&utm_emcid=test&utm_emmid=test&utm_id=test&utm_keyword=test&utm_medium=test&utm_name=test&utm_place=test&utm_product=test&utm_pubreferrer=test&utm_reader=test&utm_referrer=test&utm_serial=test&utm_session=test&utm_siteid=test&utm_social=test&utm_social-type=test&utm_source=test&utm_supplier=test&utm_swu=test&utm_term=test&utm_umguk=test&utm_userid=test&utm_viz_id=test&gbraid=test&wbraid=test&gclsrc=test&gclid=test&usqp=test&dpg_source=test&dpg_campaign=test&dpg_medium=test&dpg_content=test&admitad_uid=test&gps_adid=test&unicorn_click_id=test&adjust_creative=test&adjust_tracker_limit=test&adjust_tracker=test&adjust_adgroup=test&adjust_campaign=test&bsft_clkid=test&bsft_eid=test&bsft_mid=test&bsft_uid=test&bsft_aaid=test&bsft_ek=test&mtm_campaign=test&mtm_cid=test&mtm_content=test&mtm_group=test&mtm_keyword=test&mtm_medium=test&mtm_placement=test&mtm_source=test&pk_campaign=test&pk_medium=test&pk_source=test&vc_lpp=test&ml_subscriber=test&ml_subscriber_hash=test&rb_clickid=test&oly_anon_id=test&oly_enc_id=test&ebisAdID=test&wickedid=test&irgwc=test&fbclid=test&adfrom=test&nx_source=test&cmpid=test&asgtbndr=test&guccounter=test&guce_referrer=test&guce_referrer_sig=test&action_object_map=test&action_ref_map=test&action_type_map=test&fb_action_ids=test&fb_action_types=test&fb_comment_id=test&fb_ref=test&fb_source=test&dclid=test`

The result will be:

`https://brave.com/?itm_source=test&itm_medium=test&itm_content=test&itm_term=test&utm_ad=test&utm_cid=test&utm_medium=test&utm_social=test&dpg_source=test&dpg_campaign=test&dpg_content=test&adjust_creative=test&mtm_campaign=test&mtm_content=test&mtm_keyword=test&mtm_source=test&pk_campaign=test&pk_medium=test&action_ref_map=test`

In comparison with uBlock, uBlock would work perfect and remove all parameters 100% (even if they were blank).

Another difference is how $removeparam works by using `$document,removeparam` vs `*$document,removeparam`, because if you add the asterisk to the rules and use the same testing URL the list of parameters will change.

The resulting URL in this case should be: `https://brave.com/?itm_source=test&itm_campaign=test&itm_content=test&utm_ad=test&utm_campaign=test&utm_content=test&utm_creative=test&utm_medium=test&utm_social=test&utm_source=test&utm_term=test&dpg_source=test&dpg_medium=test&dpg_content=test&mtm_cid=test&mtm_keyword=test&pk_medium=test&pk_source=test&action_ref_map=test`

And this is the result comparing both URLs (hope it is similar to your side)

| Not removed by rules with `*$` | removed by rules with `*$` | not removed by either `$` or `*$` |
| ------------------- | --------------- | --------------------------------- |
| itm_campaign        | itm_medium      | itm_content                       |
| utm_campaign        | itm_term        | utm_ad                            |
| utm_content         | utm_cid         | utm_medium                        |
| utm_creative        | dpg_campaign    | utm_social                        |
| utm_source          | adjust_creative | dpg_source                        |
| utm_term            | mtm_campaign    | dpg_content                       |
| dpg_medium          | mtm_content     | mtm_keyword                       |
| mtm_cid             | mtm_source      | pk_medium                         |
| pk_source           | pk_campaign     | action_ref_map                    |

As you can see all the unremoved parameters have the underscore, it didn't happen in every parameter with an underscore but maybe it has to do with the fact that they are similar??, like they all have `_source`, `_content`, `_medium`, `_campaign`, `_cid`, `_creative`, and while some seem to be unique like `_keyword`, other parameter was removed with _keyword on it = `utm_keyword`, same with the one with `_ad`, `_ref` and and `utm_social` (utm_social-type)

So it is clear Brave is having trouble with it and other similar rules after the underscore, for whatever reason it is getting confused and not doing the job properly.

Of course, if you remove the underscore from the rules and the URL, and go to the URL you will notice all the parameters disappear as they are meant to.

Hope this helps to improve things, I had to edit this issue because I finally found the common factor to make it clearer and better to test it and explain the issue.
So hope this is good enough! and it can be useful to find the problem and improve the $removeparam support in Brave.

Thank you and have a nice day!

**Comment by antonok-edm** [comment]
Thanks for the detailed investigation @Emi-TheDhamphirInLoveUnderTheFrozenStar! I was able to confirm it's an issue with `adblock-rust` using your test case:

```rust
#[test]
fn test_removeparam_same_tokens() {
    let filters = vec![
        String::from("$removeparam=example1_"),
        String::from("$removeparam=example1-"),
    ];

    let (network_filters, _) = parse_filters(&filters, true, Default::default());

    let blocker_options = BlockerOptions {
        enable_optimizations: true,
    };

    let blocker = Blocker::new(network_filters, &blocker_options);

    let result = blocker.check(&Request::from_urls("https://example.com?example1_=1&example1-=2", "https://example.com", "script").unwrap());
    assert_eq!(result.rewritten_url, Some("https://example.com".into()));
    assert!(!result.matched);
}
```

**Comment by antonok-edm** [comment]
This appears to be related to the filter fusion optimizations - `$removeparam=example1_` and `$removeparam=example1-` will get fused into a single filter, but the removeparam option is not handled. I think the proper fix is to ignore these filters when considering possible fusions.

**Comment by antonok-edm** [comment]
Should be fixed in e990551
